cf-net/cf-net.c

```diff
@@ -0,0 +1,718 @@
+        }
+        else
+        {
+            local_file = xstrdup(remote_file);
+        }
+    }
+
+    if (specified_path && strcmp(local_file, "-") == 0)
+    {
+         // TODO: Should rewrite CopyRegularFileNet etc. to take fd argument
+         // and a simple helper function to open a file as well.
+        printf("Output to stdout not yet implemented (TODO)");
+        free(local_file);
+        return -1;
+    }
+
+    struct stat sb;
+    int r = cf_remote_stat(conn, true, remote_file, &sb, "file");
+    if (r!= 0)
+    {
+        printf("Could not stat: '%s'.\n", remote_file);
+    }
+    else
+    {
+        CopyRegularFileNet(remote_file, local_file, sb.st_size, true, conn);
+    }
+    free(local_file);
+    DisconnectServer(conn);
+    return 0;
+}
+
+void PrintDirs(const Item *list)
+{
+    const Item *ip = list;
+
+    while (ip != NULL)
+    {
+        // TODO: I hate this:
+        struct dirent *de = (struct dirent *) ip->name;
+        printf("%s\n", de->d_name);
+
+        ip = ip->next;
+    }
+}
+
+int CFNetOpenDir(CFNetOptions *opts, const char* hostname, char **args)
+{
+    assert(opts);
+    assert(hostname);
+    assert(args);
+    AgentConnection *conn = CFNetOpenConnection(hostname);
+    if (conn == NULL)
+    {
+        return 1;
+    }
+
+    // TODO: Propagate argv and argc from main()
+    int argc = 0;
+    while (args[argc] != NULL)
+    {
+        ++argc;
+    }
+    if (argc <= 1)
+    {
+        return invalid_command("opendir");
+    }
+
+    const char *remote_path = args[1];
+
+    Item *items = RemoteDirList(remote_path, false, conn);
+    PrintDirs(items);
+    DisconnectServer(conn);
+    return 0;
+}
+
+int CFNetMulti(const char *server)
+{
+    time_t start;
+    time(&start);
+
+    int ret = 0;
+    int attempts = 0;
+
+    printf("Connecting repeatedly to '%s' without handshakes.\n",server);
+    attempts = 0;
+
+    char *buf = xstrdup(server);
+    char *host, *port;
+    ParseHostPort(buf, &host, &port);
+    if (port == NULL)
+    {
+        port = CFENGINE_PORT_STR;
+    }
+
+    ret = 0;
+    while(ret != 1)
+    {
+        ret = JustConnect(host, port);
+        attempts++;
+    }
+
+    free(buf);
+    printf("Server unavailable after %i attempts.\n", attempts);
+
+    time_t stop;
+    time(&stop);
+    double seconds = difftime(stop,start);
+    printf("%s(): '%s' unavailable after %.f seconds (%i attempts).\n", __FUNCTION__, server, seconds, attempts);
+    return 0;
+}
+
+int CFNetMultiTLS(const char *server)
+{
+    time_t start;
+    time(&start);
+
+    printf("Multiple handshakes to '%s'.\n", server);
+    int ret = 0;
+    int attempts = 0;
+    while(ret == 0)
+    {
+        ret = CFNetConnectSingle(server, false);
+        ++attempts;
+    }
+    time_t stop;
+    time(&stop);
+
+    double seconds = difftime(stop,start);
+    printf("%s(): '%s' unavailable after %.f seconds (%i attempts).\n", __FUNCTION__, server, seconds, attempts);
```

Comment:
`__FUNCTION__` is non-standard. But anyway this message is aimed to the user, not to the developer, so it C code function name should be removed.

Applies in other places with `__FUNCTION__`.